### PR TITLE
feat(eslint-plugin): add default-param-last rule

### LINKED ISSUE #1407: [default-param-last] False positive for optional arguments denoted by `?`
**Repro**

```JSON
{
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended"
  ]
}
```

```TS
function foo(a: string, b = 'b', c?: string) {}
```

**Expected Result**

The code should pass linting, as `c?` is equivalent to `c = undefined`.

**Actual Result**

`b` fails `default-param-last`, as it is not the last parameter.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.14.0` |
| `@typescript-eslint/parser`        | `2.14.0` |
| `TypeScript`                       | `3.7.3` |
| `ESLint`                           | `6.7.2` |

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #1407

### CODE REVIEW

File: packages/eslint-plugin/src/rules/default-param-last.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        ) {
+          context.report({ node: param, messageId: 'shouldBeLast' });
+        }
+      }
+    }
+
+    return {
+      ArrowFunctionExpression: checkDefaultParamLast,
+      FunctionDeclaration: checkDefaultParamLast,
+      FunctionExpression: checkDefaultParamLast,
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/default-param-last.ts]: what do you think about adding support for other function like nodes?

- TSEmptyBodyFunctionExpression
- TSDeclareFunction
- TSConstructorType
- TSConstructSignatureDeclaration
- TSFunctionType
- TSCallSignatureDeclaration

File: packages/eslint-plugin/src/rules/default-param-last.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  defaultOptions: [],
+  create(context) {
+    /**
+     * checks if node is optional parameter
+     * @param {ASTNode} node node to evaluate
+     * @returns {boolean}
+     * @private
+     */
+    function isOptionalParam(node: TSESTree.Node): boolean {
+      return node.type === AST_NODE_TYPES.Identifier && node.optional === true;
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/default-param-last.ts]: you should check other types to: (this is producing ast)

```ts
// ObjectPattern
function foo ({ x }?: string) { }
```
```ts
// ArrayPattern
function foo ([ x ]?: string) { }
```

File: packages/eslint-plugin/docs/rules/default-param-last.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+```
+
+Examples of **correct** code for this rule:
+
+```ts
+/* eslint @typescript-eslint/default-param-last: ["error"] */
+
+function f(a: number, b = 0) {}
+function f(a: number, b?: number) {}
+function f(a: number, b?: number, c = 0) {}
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/default-param-last.md]: nit - should add the example from the issue which started it all
```suggestion
function f(a: number, b?: number, c = 0) {}
function f(a: number, b = 0, c?: number) {}
```

File: packages/eslint-plugin/src/rules/default-param-last.ts (hunk 1 above)
Comment (bradzacher) [reply packages/eslint-plugin/src/rules/default-param-last.ts]: There's no need beacuse you can't have default params in function type signatures.
Error 2371 "A parameter initializer is only allowed in a function or constructor implementation."

**Comment by a-tarasyuk** [comment]
@bradzacher @armano2 Thanks for the feedback.

**Comment by buschtoens** [comment]
omg thanks! 😍

**Pull request merged**